pkg/agent/manager/sync.go

```diff
@@ -6,189 +6,148 @@ import (
+	// in this interval.
+	//
+	// the values in `update` now belong to the cache. DO NOT MODIFY.
+	var csrs []csrRequest
+	var expiring int
+	m.cache.Update(update, func(entry *common.RegistrationEntry, svid *cache.X509SVID) {
+		var expiresAt time.Time
+		switch {
+		case svid == nil:
+			// no SVID
+		case len(svid.Chain) == 0:
+			// SVID has an empty chain. this is not expected to happen.
+		case isSVIDStale(m.c.Clk.Now(), svid.Chain[0]):
+			// SVID has expired
+			expiresAt = svid.Chain[0].NotAfter
+			expiring++
+		default:
+			// SVID is good
+			return
+		}
+		// we've exceeded the CSR limit, don't make any more CSRs
+		if len(csrs) < node.CSRLimit {
+			csrs = append(csrs, csrRequest{
+				EntryID:   entry.EntryId,
+				SpiffeID:  entry.SpiffeId,
+				ExpiresAt: expiresAt,
+			})
+		}
+	})
+	m.c.Metrics.AddSample([]string{"cache_manager", "expiring_svids"}, float32(expiring))
 
-	err = m.processEntryRequests(ctx, cEntryRequests)
-	if err != nil {
-		return err
+	if len(csrs) > 0 {
+		update, err := m.fetchUpdates(ctx, csrs)
+		if err != nil {
+			return err
+		}
+		// the values in `update` now belong to the cache. DO NOT MODIFY.
+		m.cache.Update(update, nil)
 	}
-
 	return nil
 }
 
-func (m *manager) fetchUpdates(ctx context.Context, entryRequests map[string]*entryRequest) (entries map[string]*common.RegistrationEntry, svids map[string]*node.X509SVID, err error) {
+func (m *manager) fetchUpdates(ctx context.Context, csrs []csrRequest) (_ *cache.CacheUpdate, err error) {
 	counter := telemetry.StartCall(m.c.Metrics, "manager", "sync", "fetch_updates")
 	defer counter.Done(&err)
-	// Put all the CSRs in an array to make just one call with all the CSRs.
-	csrs := [][]byte{}
-	if entryRequests != nil {
-		for _, entryRequest := range entryRequests {
-			m.c.Log.Debugf("Requesting SVID for %v", entryRequest.entry.RegistrationEntry.SpiffeId)
-			counter.AddLabel("spiffe_id", entryRequest.entry.RegistrationEntry.SpiffeId)
-			csrs = append(csrs, entryRequest.CSR)
+
+	req := &node.FetchX509SVIDRequest{}
+
+	// TODO: The node API is currently insufficient to handle multiple
+	// registration entries mapping to the same SPIFFE ID (since results are
+	// keyed by SPIFFE ID). This code will use the same SVID for any
+	// registration entry sharing a SPIFFE ID and should be fixed when
+	// the node API is fixed.
+	privateKeys := make(map[string]*ecdsa.PrivateKey, len(csrs))
+	for _, csr := range csrs {
+		log := m.c.Log.WithField("spiffe_id", csr.SpiffeID)
+		if !csr.ExpiresAt.IsZero() {
+			log = log.WithField("expires_at", csr.ExpiresAt.Format(time.RFC3339))
 		}
-	}
+		log.Info("Renewing X509-SVID")
+		counter.AddLabel("spiffe_id", csr.SpiffeID)
 
-	update, err := m.client.FetchUpdates(ctx, &node.FetchX509SVIDRequest{Csrs: csrs})
-	if err != nil {
-		return nil, nil, err
-	}
+		// Skip CSR for the same SPIFFE ID... for now (see above TODO)
+		if _, ok := privateKeys[csr.SpiffeID]; ok {
+			continue
```

Comment:
I know it will change in a future, but  maybe we can add a log to notify that private key will no be used, just to be able to debug what is happening because we "lost" some keys

Reply:
sure, i will add a debug log.